lxd/cluster/gateway.go

```diff
@@ -177,25 +177,28 @@ func (g *Gateway) HandlerFuncs(nodeRefreshTask func(*APIHeartbeat)) map[string]h
 				return
 			}
 
-			nodes := make([]db.RaftNode, 0)
-			for _, node := range heartbeatData.Members {
-				if node.Raft {
-					nodes = append(nodes, db.RaftNode{
-						ID:      node.ID,
+			raftNodes := make([]db.RaftNode, 0)
+			for _, node := range heartbeatData.Nodes {
+				if node.RaftID > 0 {
+					raftNodes = append(raftNodes, db.RaftNode{
+						ID:      node.RaftID,
 						Address: node.Address,
 					})
 				}
 			}
 
-			// Accept Raft node updates from any node (joining nodes just send raft nodes heartbeat data).
-			logger.Debugf("Replace current raft nodes with %+v", nodes)
-			err = g.db.Transaction(func(tx *db.NodeTx) error {
-				return tx.RaftNodesReplace(nodes)
-			})
-			if err != nil {
-				logger.Errorf("Error updating raft nodes: %v", err)
-				http.Error(w, "500 failed to update raft nodes", http.StatusInternalServerError)
-				return
+			// Check we have been sent at least 1 raft node before wiping our set.
+			if len(raftNodes) > 0 {
```

Comment:
@freeekanayaka @stgraber I added this as a safety check to ensure that if for any reason there are no raft nodes sent in heartbeat, then it wont replace the existing local set. That way if a new node receives a heartbeat in the old format it will wait until the leader is upgraded. Does this sound reasonable?

Reply:
This is probably a matter of programming style. In general I'd expect whoever send the heartbeat request to be (or believe it is) the leader and therefore have a non-empty raft_nodes table. If this assumption is correct, then I'd put a panic here, to not silently swallow something that might be an indication of a broader problem. YMMV

Reply:
@freeekanayaka I'm happy to add an error message, thats a good point, although I think panicing and crashing the node would cause more issues as it would prevent the nodeRefresh task from running (which handles the upgrades). https://github.com/lxc/lxd/blob/072b277de0dda9b9dadc8707f11ad87d5c74ddb3/lxd/cluster/gateway.go#L212 here https://github.com/lxc/lxd/blob/bbaacfb4b994dcb25ed50b3c41241b1dd8577615/lxd/daemon.go#L1350-L1357

Reply:
@freeekanayaka added error message.